dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/TranslatableObject.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2004-2004-2021, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.common;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.commons.lang3.StringUtils;
+import org.hisp.dhis.schema.annotation.Gist;
+import org.hisp.dhis.translation.Translation;
+import org.hisp.dhis.user.UserSettingKey;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper;
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
+
+/**
+ * Base class for translatable object.
+ * <p>
+ * Contains necessary methods for storing and retrieving translation values.
+ * <p>
+ * Beside extending this object, translatable object needs to have a column
+ * translations with type jblTranslations defined in hibernate mapping file.
+ */
+public class TranslatableObject
+{
+    /**
+     * Set of available object translation, normally filtered by locale.
+     */
+    protected Set<Translation> translations = new HashSet<>();
+
+    /**
+     * Cache for object translations, where the cache key is a combination of
+     * locale and translation property, and value is the translated value.
+     */
+    protected Map<String, String> translationCache = new HashMap<>();
```

Comment:
I guess the `TranslatableObject` instance could come from hibernate cache which means there should be a chance this field is accessed by multiple threads. I'd go with a `ConcurrentHashMap` and make the field `final` to be safe.